c-api/src/rewriter.rs

```diff
@@ -52,37 +56,45 @@ pub extern "C" fn lol_html_rewriter_build(
     };
 
     let output_sink = ExternOutputSink::new(output_sink, output_sink_user_data);
-    let rewriter = unwrap_or_ret_null! { HtmlRewriter::try_new(settings, output_sink) };
+    let rewriter = unwrap_or_ret_null! { lol_html::HtmlRewriter::try_new(settings, output_sink) };
 
-    to_ptr_mut(rewriter)
+    to_ptr_mut(HtmlRewriter(Some(rewriter)))
 }
 
 #[no_mangle]
 pub extern "C" fn lol_html_rewriter_write(
-    rewriter: *mut HtmlRewriter<'static, ExternOutputSink>,
+    rewriter: *mut HtmlRewriter,
     chunk: *const c_char,
     chunk_len: size_t,
 ) -> c_int {
     let chunk = to_bytes!(chunk, chunk_len);
-    let rewriter = to_ref_mut!(rewriter);
+    let rewriter = to_ref_mut!(rewriter)
+        .0
+        .as_mut()
+        .expect("cannot call `lol_html_rewriter_write` after calling `end()`");
 
     unwrap_or_ret_err_code! { rewriter.write(chunk) };
 
     0
 }
 
 #[no_mangle]
-pub extern "C" fn lol_html_rewriter_end(
-    rewriter: *mut HtmlRewriter<'static, ExternOutputSink>,
-) -> c_int {
-    let rewriter = to_ref_mut!(rewriter);
+pub extern "C" fn lol_html_rewriter_end(rewriter: *mut HtmlRewriter) -> c_int {
+    let rewriter = to_ref_mut!(rewriter)
+        .0
+        .take() // Using `take()` allows calling `free()` afterwards (it will be a no-op).
+        .expect("cannot call `lol_html_rewriter_end` after calling `end()`");
```

Comment:
I guess it should be "after calling `lol_html_rewriter_free` or `lol_html_rewriter_end`"?

Reply:
If this is called after free(), then one of two things happened:

1. `rewriter.0` was `None` when dropped, so `end()` was called in the past, so the error message is correct (free() is irrelevant to the panic).
2. `rewriter.0` was `Some` when dropped, so this value has already been dropped, so this is UB. In practice, I think the bytes will not be zeroed, so this panic won't execute.

In either case, I don't think there's a need to change the message.